pkg/crypto/protocol/difficulty.go

```diff
@@ -0,0 +1,17 @@
+package protocol
+
+import (
+	"encoding/binary"
+	"math/bits"
+)
+
+// CountHashDifficultyBits returns the number of leading zero bits in the given byte
+// slice. It returns an error if the byte slice is all zero bits.
+func CountHashDifficultyBits(bz [32]byte) int {
```

Comment:
```suggestion
// TODO_MAINNET: Consider generalizing difficulty to a target hash.
// https://bitcoin.stackexchange.com/questions/107976/bitcoin-difficulty-why-leading-0s
https://bitcoin.stackexchange.com/questions/121920/is-it-always-possible-to-find-a-number-whose-hash-starts-with-a-certain-number-o
func CountHashDifficultyBits(bz [32]byte) int {
```

Reply:
@Olshansk Here's what it would look like if we made it into a value comparison:

```go
// HasMinDifficultyBits returns the number of leading zero bits in the given byte
// slice. It returns an error if the byte slice is all zero bits.
func HasMinDifficultyBits(bz [32]byte, minDifficultyBits uint64) bool {
	if minDifficultyBits == 0 {
		return true
	}

	var mask [32]byte
	for i := uint64(0); i < 32; i++ {
		if minDifficultyBits < 8 {
			mask[i] = 0xFF << (8 - minDifficultyBits)
			break
		}

		if i < minDifficultyBits/8 {
			mask[i] = 0xFF
		}
	}
	comp := bytes.Compare(mask[:], bz[:])

	return comp == 1
}
```

I take the point that there are other ways to measure difficulty than to count zero bits. However, I find this much less readable and intuitive by comparison.

Reply:
Since merging, I've read through `relay_mining_difficulty.proto` and `update_relay_mining_difficulty.go` and had [a conversation with chatGPT](https://chatgpt.com/share/303403bc-75aa-419a-bc7f-85496e0684b8). I had previously missed the fact about more granularity of control, mistakenly thinking that one bit at a time is as granular as you can get. :sweat_smile: